src/composables/storageRef.ts

```diff
@@ -51,20 +62,66 @@ export function useStorageRef<T = string | object | any[]>(
   }
 
   function setStorageState(val: T | null) {
-    WalletStorage.set(storageKey, (val && serializer?.write) ? serializer.write(val) : val);
+    storage?.set(storageKey, (val && serializer?.write) ? serializer.write(val) : val);
   }
 
   // Restore state and run watchers
   (async () => {
-    let restoredValue = await WalletStorage.get<T | null>(storageKey);
+    if (isSecure) {
+      // Check if value exists in insecure storage and migrate it to secure storage
+      let unencryptedValue = await WalletStorage.get<T | null>(storageKey);
+      if (storageKey === STORAGE_KEYS.mnemonic) {
+        unencryptedValue = await migrateMnemonicToSecureStorage(unencryptedValue!);
+      }
+      if (unencryptedValue !== null) {
+        onRestored?.(unencryptedValue);
+        setLocalState(unencryptedValue);
+      }
+
+      // Create secure storage after we've migrated the mnemonic and set the password
+      const { secureStorage, isLoggedIn } = useSecureStorage();
+      await watchUntilTruthy(secureStorage);
+      storage = secureStorage.value!;
+
+      // Move the unencrypted value to secure storage and remove it from insecure storage
+      if (IS_MOBILE_APP && unencryptedValue !== null) {
+        setStorageState(unencryptedValue);
+        WalletStorage.remove(storageKey);
+      }
+
+      /**
+       * Clear the state when the user logs out.
+       * This ensures that the state is not leaked if someone removes the modal from the DOM.
+       *
+       * TODO This breaks the UI behind the modal while user is logged out
+       * because the state is not restored yet.
+       */
+      watch(isLoggedIn, async (val) => {
```

Comment:
I strongly think that this composable should not have any logic connected to user's logged in state.